timefold-solver/src/main/java/com/baeldung/timefoldsolver/ShiftScheduleConstraintProvider.java

```diff
@@ -0,0 +1,37 @@
+package com.baeldung.timefoldsolver;
+
+import ai.timefold.solver.core.api.score.buildin.hardsoft.HardSoftScore;
+import ai.timefold.solver.core.api.score.stream.Constraint;
+import ai.timefold.solver.core.api.score.stream.ConstraintFactory;
+import ai.timefold.solver.core.api.score.stream.ConstraintProvider;
+
+import static ai.timefold.solver.core.api.score.stream.Joiners.equal;
+
+public class ShiftScheduleConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+                atMostOneShiftPerDay(constraintFactory),
+                requiredSkill(constraintFactory)
+        };
+    }
+
+    public Constraint atMostOneShiftPerDay(ConstraintFactory constraintFactory) {
+        return constraintFactory.forEach(Shift.class)
+                .join(Shift.class,
+                        equal(shift -> shift.getStart().toLocalDate()),
+                        equal(Shift::getEmployee))
+                .filter((shift1, shift2) -> shift1 != shift2)
```

Comment:
Is it safe to use reference equality? I'm guessing this means that ConstraintFactory guarantees it will use the exact same instance of a particular shift repeatedly

Reply:
Timefold Solver internally guarantees that 1 solution with one set of shifts will be used in 1 score director (= repeated score calculations). This mean that something like `same(Shift::getEmployee)` would work (if such an operation would exist). But `toLocalDate()` doesn't guarantee the same instance on each call, so it's better to use equals semantics.

Reply:
It's unsafe to use equality that depends on the entire state of the object if the object is a planning entity.
So something like `equal(Shift::getDependentShift)` with Shift doing equality based on `Shift.employee` too (which changes during planning) would cause havoc. But Shift doing equality based on a ID, or even start/end datetimes is fine.

Reply:
Generally speaking, this just works. Almost no users run into trouble with how it works out of the box. So I wouldn't go into trying to explain the nitty gritty in the article here.

Reply:
So to summarise, are we saying that:
1. We can rely on the base PlanningEntity (Shift in this case) as being reference-stable for the purpose of these APIs
2. But we cannot rely on any PlanningEntity fields (e.g. nested Shift, or reference to another PlanningEntity) to be reference-stable

The reason that 1. was confusing to me is that, in the wider Java world, it's very unusual to see an intentional use of reference equality for non primitive types. Usually a line like this would be an unintentional bug, and it's only because you explicitly designed for it that it's ok.

2. Makes sense, I suppose this means that one limitation of the solver is that you can't have constraints that create cyclic dependencies with the rest of the solver state, if I'm understanding correctly. And instead we impose constraints that partition the solution space, but do not introduce cycles. Let me know if I'm misunderstanding anything here.

> Generally speaking, this just works. Almost no users run into trouble with how it works out of the box. So I wouldn't go into trying to explain the nitty gritty in the article here.

Yes I agree not to clutter the article with these details, just making sure I'm understanding what's going on exactly.